### ISSUE TITLE
Add editorConfig preset to package.json?

### ISSUE THREAD

**Issue by kevinSuttle** [editorconfig/editorconfig-core-js#30]
In the spirit of ESLint and Babel, would it make sense to allow editorConfig settings in package.json? 
I feel like it probably doesn't, but I wanted to have it documented either way to be definitive.

**Comment by jedmao** [comment]
That's rather JavaScript specific, which isn't really in the design goals of EditorConfig.

**Comment by kevinSuttle** [comment]
@jedmao No. It's not. 

Look at how YAML does it. 
https://docs.cloudfoundry.org/devguide/deploy-apps/manifest.html#multi-manifests

**Comment by jedmao** [comment]
Ah, see, I didn't realize this was on the core-js repo. Seeing as it is, I have no problems with such a feature; though, we still need to look in subfolders for both `.editorconfig` and `package.json` files, which means the `package.json` file needs to define `"root": true` if that's the case. Then, what happens if you have both an `.editorconfig` as well as a `package.json` file in the same folder, both with an EditorConfig configuration? Which one takes precedent? I can see some issues to address here.

**Comment by kevinSuttle** [comment]
Tools like eslint have this exact scenario. It's just a priority order. http://eslint.org/docs/user-guide/configuring#configuration-file-formats

**Comment by corysimmons** [comment]
Would prefer to see someone make some standard config file (or folder) for all configs to live in rather than continuing to clutter package.json with xyz.

...that said, if it does get in, I'll probably use it.

**Comment by jonathantneal** [comment]
Likewise, I will use this. @jedmao, what can we do to help? Do we need to look into the plugin and make a PR for the functionality? Could someone help with documentation? I’d love to see this get in on the next release.

**Comment by jedmao** [comment]
First, we need to come up with an acceptable spec. At first glance, I'm imagining something like the following:
```json
{
  "editorConfig": {
    "root": true,
    "settings": [
      {
        "*": {
          "indent_style": "space",
          "indent_size": 4
        }
      },
      {
        "package.json": {
          "indent_size": 2
        }
      }
    ]
  }
}
```
It doesn't look as clean as the INI file format that `.editorconfig` uses, but this is the only way I could imagine supporting it in a `package.json` file.

BTW – you should theoretically be able to have any number of `package.json` files throughout your project. I would imagine that if both `package.json` and `.editorconfig` files were in a directory that the `.editorconfig` file would take precedence, but I'm open to discussion.

I took a look at the [`multilevel-ini`](https://www.npmjs.com/package/multilevel-ini) and other INI parsing packages, but none of them account for ordered settings. I think the example structure I provided above is the only structure that would support this.

**Comment by jedmao** [comment]
Another alternative would be the following structure:
```json
{
  "editorConfig": [
    ["root", true],
    ["*", {
      "indent_style": "space",
      "indent_size": 4
    }]
  ]
}
```

**Comment by jonathantneal** [comment]
@jedmao, thanks for the fast reply!

I’m a fan of the first syntax you’ve shared, and I would expect, as you’ve said, that the `.editorconfig` file takes precedent when both it and a `package.json` are present. Would this support an `extends` property being added alongside `root` and the other rules?

**Comment by jedmao** [comment]
@jonathantneal anything above existing functionality would be a separate ticket. Please open another issue for that feature request and detail how it might work.

BTW – I'm not at all suggesting I have any bandwidth to implement this `package.json` feature. I still have zero interest in the feature for my own use, so I won't be using it.

I would, however, consider merging in a PR, if it were implemented properly. That is, of course, the purpose of this discussion, to talk about what a proper implementation would be.

**Comment by kevinSuttle** [comment]
@jonathantneal See https://github.com/editorconfig/editorconfig/issues/245

**Comment by jonathantneal** [comment]
Hey there. What can I do to help with moving this forward? Do we still need help defining the request more clearly? We have a year of sharable configs under our belt for most if-not-all linters. Does the configuration just need an outlined JSON version? Is this something we should be writing plugins for Atom / Sublime first for?

**Comment by ButuzGOL** [comment]
+

**Comment by andreiglingeanu** [comment]
I would love to contribute somehow on getting this thing moving! There's really no other way to re-use `.editorconfig`'s across multiple repos.

**Comment by jedmao** [comment]
Nope. No other way. And remember, this is just for the js core. Other cores would need some kind of implementation too. I think we should stay focused on the OP's request and just support configuration in a package.json file for now and then start worrying about how the "extends" feature would work.

**Comment by andreiglingeanu** [comment]
@jedmao Every _core_ has its own options parsing code duplicated? That is, `.editorconfig` is pretty much hardcoded in each of them, right?

**Comment by jedmao** [comment]
Yes, it has to be duplicated because the cores are written in different languages. The only thing that's not duplicated are the tests, which are shared across cores.

**Comment by andreiglingeanu** [comment]
Seems reasonable enough, thanks.

**Comment by jedmao** [comment]
I just did a pretty hefty rewrite of the source code into TypeScript. Anyone interested in taking this feature further? Been a bit of silence.

**Comment by andreiglingeanu** [comment]
I'd love to hop in to help a bit but I'm not really able to get the local dev setup working. The local `bin/editorconfig` definitely doesn't get created after doing `npm install; npm link`. Is that something easily fixable?

**Comment by jedmao** [comment]
@andreiglingeanu I think the issue you are experiencing is related to https://github.com/editorconfig/editorconfig-core-js/issues/55. Perhaps we should do it the other way and remove the `./bin` folder from package.json and restore the `./bin/editorconfig`.

**Comment by andreiglingeanu** [comment]
Hey @jedmao, yes, that it is. The `bin/editorconfig` file did not get generated at all in my local setup.

**Comment by jedmao** [comment]
@andreiglingeanu, aha! See #60. Now that the `package.json` file expects to be published in the `./dist` folder, that's where the `npm link` command needs to be run. Make sense?

**Comment by andreiglingeanu** [comment]
We are good to go! Expect the pull request soon.

**Comment by BtM909** [comment]
Maybe I'm late to the party, but would this adhere to the [cosmiconfig configuration file spec](https://github.com/davidtheclark/cosmiconfig)?

One of the main reasons to avoid a mandatory "root": true would be that it 
> Stops at the first configuration found, instead of finding all that can be found up the directory tree and merging them automatically.

Secondly, it would make it easier to adapt the other (defacto) standard to support different types of configuration files. 

Thirdly, priority of configuration files can also be configured (and therefore predictable).

Would Editorconfig be open for other configuration file formats? Or would it be easier to open a separate issue to discuss that?

**Comment by jedmao** [comment]
@BtM909 I was looking at [`package-json`](https://www.npmjs.com/package/package-json) myself, but it looks like [`cosmiconfig`](https://github.com/davidtheclark/cosmiconfig) has a bit more popularity. Not sure the delta between them.

As for other configuration file formats. What did you have in mind?

**Comment by BtM909** [comment]
Sorry, completely missed your update... My apologies!

> As for other configuration file formats. What did you have in mind?
What formats are you specifically referring to? I was just references the fact that cosmiconfig supports multiple file formats by default, like:

- a package.json property
- a JSON or YAML, extensionless "rc file"
- an "rc file" with the extensions .json, .yaml, .yml, or .js.
- a .config.js CommonJS module

**Comment by mattgreenfield** [comment]
@andreiglingeanu  - did you ever make the pull request for this? Thanks for your help.

**Comment by ColdHandz** [comment]
Is there any movement on this feature?

**Comment by lucasvazq** [comment]
bump

**Comment by SrBrahma** [comment]
Up! I hate small files cluttering my workspace. It could really be in the package.json.

**Comment by ColtHands** [comment]
@lucasvazq @SrBrahma

Its not hard to implement this feature, but the main problem is that this repo is long dead, it uses really old versions and heck even `tslint` which has been long dead as well.

Only way for me to implement a feature is to fork it and create a new package, even if the code itself would be easy to implement, I'd be stuck with version bumping, security issues and manual version testing to make sure it all works.

If you like the package you could fork it and create your own version.

### But i hightly suggest you use something like
`prettier` or `eslint` for your tabs and spaces

for end of line you could just use your editors features or just edit it directly with git
`git config --global --edit`
and set your end of line there

**Comment by hildjj** [comment]
OK, we're back up and running.  I'm looking through open issues before doing a release.  Does anyone still want this?

**Comment by jrnewton** [comment]
> Does anyone still want this?

Yes please!

**Comment by jonathantneal** [comment]
Likewise, I will use this.

**Comment by HolgerJeromin** [comment]
> This approach also means that we can't have two rule sections that have the same pattern, but that seems like a good restriction anyway.

We also have no rule order. I am not sure if this is important but I have this scenario in mind:
```json
{
  "*": {"optionA": true},
  "subfolder/*": {"optionA": false}
}

**Comment by hildjj** [comment]
Given that order needs to matter, ParseStringResult is probably the correct type.  Its typescript definition is:

```typescript
export type SectionName = string | null
export interface SectionBody { [key: string]: string }
export type ParseStringResult = [SectionName, SectionBody][]
```

**Comment by Lioness100** [comment]
Is there any way shareable configs could be added to this? (like: `"editorConfig" : { "extends": ["some-package", "./local.json"] }`)

**Comment by hildjj** [comment]
Shareable configs are a separate feature request that needs to change the [specification](https://github.com/editorconfig/specification).
